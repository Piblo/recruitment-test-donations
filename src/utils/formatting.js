import { timeFormat } from 'd3';

export function formatDate(date) {
  const d = date instanceof Date
    ? date
    : new Date(date);

  return timeFormat('%d/%m/%Y')(d);
}
