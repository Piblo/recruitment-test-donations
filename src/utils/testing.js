import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { theme } from '../components';

export function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>
      {component}
    </ThemeProvider>
  );
}
