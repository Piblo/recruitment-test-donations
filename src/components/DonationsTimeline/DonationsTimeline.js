import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { Avatar, Container } from '..';
import { formatDate, getCurrencySymbol } from '../../utils/formatting';

const DonationsTimeline = ({ donations }) => (
  <React.Fragment>
    <Container direction="row" alignItems="center">
      {donations.map((donation, index) =>
        <Container key={index} alignItems="center" grow={1}>
          <Avatar src={donation.imageUrl} />
          <DonorName>{donation.donorDisplayName}</DonorName>
          {/* <MessageContainer><Message>{donation.message}</Message></MessageContainer> */}
          <Amount>{getCurrencySymbol(donation.currencyCode) + donation.amount}</Amount>
          <VerticalLine />
          <EventCircle />
          <DonationDate>{formatDate(donation.donationDate)}</DonationDate>
        </Container>
      )}
    </Container>
    <Timeline />
  </React.Fragment>
);

export default DonationsTimeline;

DonationsTimeline.propTypes = {
  donations: PropTypes.array
};

const Timeline = styled.div`
  background-color: ${props => props.theme.palette.primary};
  height: 3px;
  transform: translateY(-50%);
`;

// const DonationContainer = styled(Container)`
//   max-width: 25%;
// `;

const DonorName = styled.p`
  margin-bottom: 6px;
`;

const Amount = styled.span`
  color: ${props => props.theme.palette.primary};
  font-size: 1.3em;
  margin-bottom: 6px;
`;

const DonationDate = styled.div`
  color: ${props => props.theme.palette.textLight};
  position: absolute;
  bottom: -32px;
  font-size: 0.8em;
`;

const VerticalLine = styled.div`
  border-left: 1px dashed #ECF0F1;
  height: 40px;
`;

const EventCircle = styled.div`
  border-radius: 50%;
  height: 15px;
  width: 15px;
  background-color: ${props => props.theme.palette.primary};
  transform: translateY(50%);
`;

const MessageContainer = styled.div`
  width: 100%;
`;

const Message = styled.p`
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  color: ${props => props.theme.palette.textLight};
  font-size: 0.8em;
`;
