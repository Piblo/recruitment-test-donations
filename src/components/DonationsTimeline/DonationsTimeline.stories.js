import React from 'react';
import { storiesOf } from '@storybook/react';
import DonationsTimeline from './DonationsTimeline';

storiesOf('DonationsTimeline', module)
  .add('default', () => (
    <DonationsTimeline donations={donations} />
  ));

const donations = [{
  amount: 3,
  currencyCode: 'GBP',
  date: new Date('2019-01-03'),
  donorName: 'Anthony (Tony) Rudd',
  avatarUrl: 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif',
  message: 'Have just had an ICD fitted, here is a small amount to help things, tis all I can afford.'
},
{
  amount: 10.5,
  currencyCode: 'GBP',
  date: new Date('2019-01-01'),
  donorName: 'Michael Gibson',
  avatarUrl: 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif',
  message: 'Every little helps!!!'
},
{
  amount: 3,
  currencyCode: 'GBP',
  date: new Date('2019-01-02'),
  donorName: 'Anthony (Tony) Rudd',
  avatarUrl: 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif',
  message: 'Have just had an ICD fitted, here is a small amount to help things, tis all I can afford.'
}];
