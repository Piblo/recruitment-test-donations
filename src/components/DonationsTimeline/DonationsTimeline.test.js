import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import DonationsTimeline from './DonationsTimeline';
import { Donation } from '..';
import { renderWithTheme } from '../../utils/testing';

describe('<DonationsTimeline>', () => {
  it('Displays donation events', () => {
    const component = shallow(<DonationsTimeline donations={donations} />);
    const sortedDonations = donations.sort((a, b) => a.date - b.date);

    sortedDonations.forEach((donation, index) => {
      const donationComponent = component.find(Donation).at(index);
      expect(donationComponent.prop('date')).toEqual(donation.date);
      expect(donationComponent.prop('amount')).toEqual(donation.amount);
      expect(donationComponent.prop('avatarUrl')).toEqual(donation.avatarUrl);
      expect(donationComponent.prop('currencyCode')).toEqual(donation.currencyCode);
      expect(donationComponent.prop('donorName')).toEqual(donation.donorName);
    });
  });

  it('Matches snapshot', () => {
    const tree = renderWithTheme(<DonationsTimeline donations={donations} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const donations = [{
  amount: 3,
  currencyCode: 'GBP',
  date: new Date('2019-01-03'),
  donorName: 'Anthony (Tony) Rudd',
  avatarUrl: 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif',
  message: 'Have just had an ICD fitted, here is a small amount to help things, tis all I can afford.'
},
{
  amount: 10.5,
  currencyCode: 'GBP',
  date: new Date('2019-01-01'),
  donorName: 'Michael Gibson',
  avatarUrl: 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif',
  message: 'Every little helps!!!'
},
{
  amount: 3,
  currencyCode: 'GBP',
  date: new Date('2019-01-02'),
  donorName: 'Anthony (Tony) Rudd',
  avatarUrl: 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif',
  message: 'Have just had an ICD fitted, here is a small amount to help things, tis all I can afford.'
}];
