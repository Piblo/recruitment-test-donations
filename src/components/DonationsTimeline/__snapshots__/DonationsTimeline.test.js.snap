// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DonationsTimeline> Matches snapshot 1`] = `
Array [
  .c2 {
  border-radius: 50%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: unset;
  -ms-flex: unset;
  flex: unset;
  -webkit-box-pack: unset;
  -webkit-justify-content: unset;
  -ms-flex-pack: unset;
  justify-content: unset;
  -webkit-flex-basis: unset;
  -ms-flex-preferred-size: unset;
  flex-basis: unset;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: unset;
  -webkit-justify-content: unset;
  -ms-flex-pack: unset;
  justify-content: unset;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  position: relative;
  position: relative;
}

.c3 {
  margin-bottom: 6px;
}

.c4 {
  color: #E887ED;
  font-size: 1.3em;
  margin-bottom: 6px;
}

.c7 {
  color: #AAAAAA;
  position: absolute;
  bottom: -32px;
  font-size: 0.8em;
}

.c5 {
  border-left: 1px dashed #ECF0F1;
  height: 40px;
}

.c6 {
  border-radius: 50%;
  height: 15px;
  width: 15px;
  background-color: #E887ED;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

<div
    className="c0"
    direction="row"
  >
    <div
      className="c1"
      direction="column"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <img
        className="c2"
        src="https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif"
      />
      <p
        className="c3"
      >
        Michael Gibson
      </p>
      <span
        className="c4"
      >
        £10.5
      </span>
      <div
        className="c5"
      />
      <div
        className="c6"
      />
      <div
        className="c7"
      >
        01/01/2019
      </div>
    </div>
    <div
      className="c1"
      direction="column"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <img
        className="c2"
        src="https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif"
      />
      <p
        className="c3"
      >
        Anthony (Tony) Rudd
      </p>
      <span
        className="c4"
      >
        £3
      </span>
      <div
        className="c5"
      />
      <div
        className="c6"
      />
      <div
        className="c7"
      >
        02/01/2019
      </div>
    </div>
    <div
      className="c1"
      direction="column"
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <img
        className="c2"
        src="https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif"
      />
      <p
        className="c3"
      >
        Anthony (Tony) Rudd
      </p>
      <span
        className="c4"
      >
        £3
      </span>
      <div
        className="c5"
      />
      <div
        className="c6"
      />
      <div
        className="c7"
      >
        03/01/2019
      </div>
    </div>
  </div>,
  .c0 {
  background-color: #E887ED;
  height: 3px;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div
    className="c0"
  />,
]
`;
